controller/array_action/array_mediator_svc.py

```diff
@@ -230,32 +278,111 @@ def create_volume(self, name, size_in_bytes, capabilities, pool, volume_prefix="
+
+    def _create_fcmap(self, source_volume_name, target_volume_name):
+        logger.info("creating FlashCopy Mapping from '{0}' to '{1}'".format(source_volume_name, target_volume_name))
+        try:
+            self.client.svctask.mkfcmap(source=source_volume_name, target=target_volume_name)
+        except (svc_errors.CommandExecutionError, CLIFailureError) as ex:
+            if not is_warning_message(ex.my_message):
+                if FCMAP_ALREADY_EXIST in ex.my_message:
+                    logger.info(("FlashCopy Mapping already exists"
+                                 " for source '{0}' and target '{1}'").format(source_volume_name,
+                                                                              target_volume_name))
+                else:
+                    raise ex
+
+    def _start_fcmap(self, fcmap_id):
+        logger.info("starting FlashCopy Mapping '{0}'".format(fcmap_id))
+        try:
+            self.client.svctask.startfcmap(prep=True, object_id=fcmap_id)
+        except (svc_errors.CommandExecutionError, CLIFailureError) as ex:
+            if not is_warning_message(ex.my_message):
+                if FCMAP_ALREADY_COPYING in ex.my_message:
+                    logger.info("FlashCopy Mapping '{0}' already copying".format(fcmap_id))
+                else:
+                    raise ex
+
+    def _create_and_start_fcmap(self, source_volume_name, target_volume_name):
+        self._create_fcmap(source_volume_name, target_volume_name)
+        target_cli_volume = self._get_cli_volume(target_volume_name)
+        self._start_fcmap(target_cli_volume.FC_id)
+        return target_cli_volume
+
+    def _delete_unstarted_fcmap(self, fcmap_id):
+        logger.info("deleting fcmap with id : {0}".format(fcmap_id))
+        try:
+            self.client.svctask.rmfcmap(object_id=fcmap_id)
+        except (svc_errors.CommandExecutionError, CLIFailureError) as ex:
+            if not is_warning_message(ex.my_message):
+                logger.warning("Failed to delete fcmap '{0}': {1}".format(fcmap_id, ex))
+
+    def _delete_target_volume_if_exist(self, target_volume_name):
+        target_cli_volume = self._get_cli_volume_if_exists(target_volume_name)
+        if target_cli_volume and target_cli_volume.FC_id:
+            self._delete_unstarted_fcmap(target_cli_volume.FC_id)
+        if target_cli_volume:
+            self._delete_volume(target_volume_name, not_exist_err=False)
+
+    def _create_snapshot(self, target_volume_name, source_volume_name):
+        self._create_similar_volume(target_volume_name, source_volume_name)
+        try:
+            return self._create_and_start_fcmap(source_volume_name, target_volume_name)
+        except (svc_errors.CommandExecutionError, CLIFailureError) as ex:
+            logger.error("Failed to create snapshot '{0}': {1}".format(target_volume_name, ex))
+            self._delete_target_volume_if_exist(target_volume_name)
+            raise ex
```

Comment:
in rollback fcmap should also be deleted and also rollback should in have retries (Eran's rrquest in meeting)

Reply:
what do you mean? I don't delete a mapped volume (I don't think it's possible), if that's what you mean

Reply:
but you should delete it.
This is rollback we talked about

Reply:
delete what?
could you list the gaps that you see in "_delete_target_volume_if_exist"?

Reply:
fcmap example should be deleted.

Reply:
as I wrote, I don't think it's possible to delete a target volume without deleting the fcmap, so I'm still not sure what you mean.
could you point me to where exactly the fcmap deletion is missing?

Reply:
The scenario is:
- create similar volume
- create fcmap
- start fcmap
When you do rollback you shoul
- delete fcmap
- delete volume

I don't see where you delete fcmap.

Reply:
this is the method implementation:
```
377 def _delete_target_volume_if_exist(self, target_volume_name):
378         target_cli_volume = self._get_cli_volume_if_exists(target_volume_name)
379         if target_cli_volume and target_cli_volume.FC_id:
380             self._delete_unstarted_fcmap(target_cli_volume.FC_id)
381        if target_cli_volume:
382             self._delete_volume_by_name(target_volume_name, not_exist_err=False)
```
is line 380 does not cover something/not clear enough?
will it be ok if I implement a retry similar to the original one? (e.g. `map_volume_by_initiators`)

Reply:
oh
S only method name is not clear enough. call it _delete_target_volumeand_fcmap_if_exist
the answer for the second question  is "yes"

Reply:
oh
method name is misleaing Call it _delete_target_volume_and_fc_map_if_exist
The answer for the second question is "yes"

Reply:
renamed to `_rollback_create_snapshot` in https://github.com/IBM/ibm-block-csi-driver/pull/171